pkg/ottl/ottlfuncs/func_replace_all_patterns.go

```diff
@@ -59,40 +59,73 @@ func replaceAllPatterns[K any](target ottl.PMapGetter[K], mode string, regexPatt
 		if err != nil {
 			return nil, err
 		}
-		if fn.IsEmpty() {
-			replacementVal, err = replacement.Get(ctx, tCtx)
-			if err != nil {
-				return nil, err
-			}
-		} else {
-			fnVal := fn.Get()
-			replacementExpr, errNew := fnVal.Get(&replaceAllPatternFuncArgs[K]{Input: replacement})
-			if errNew != nil {
-				return nil, errNew
-			}
-			replacementValRaw, errNew := replacementExpr.Eval(ctx, tCtx)
-			if errNew != nil {
-				return nil, errNew
-			}
-			replacementValStr, ok := replacementValRaw.(string)
-			if !ok {
-				return nil, fmt.Errorf("replacement value is not a string")
-			}
-			replacementVal = replacementValStr
+		replacementVal, err = replacement.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
 		}
 		updated := pcommon.NewMap()
 		updated.EnsureCapacity(val.Len())
 		val.Range(func(key string, originalValue pcommon.Value) bool {
 			switch mode {
 			case modeValue:
 				if compiledPattern.MatchString(originalValue.Str()) {
+					if !fn.IsEmpty() {
+						result := []byte{}
+						submatches := compiledPattern.FindStringSubmatchIndex(originalValue.Str())
```

Comment:
Can we use `FindAllStringSubmatchIndex` instead? I believe this one doesn't return additional matches if the input string contains newlines. We're already using the `All` "function type" (the meanings behind the names are [documented here](https://pkg.go.dev/regexp#pkg-overview)) for `ReplaceAllString`, so I think we should match that here.

Reply:
My understand here is that we only care about the leftmost submatch.

For example, in the string "goodbye world and world" if we use the regex - "(world)"

we would find the leftmost submatch "world" and the result would be
"goodbye hash(world) and hash(world)"

if we use `FindAllStringSubmatchIndex` (returns [][]string] we need to loop through it and end up with something like "goodbye hash(worldworld) and hash(worldworld)" because all the submatches get appended.

Reply:
What about for the case where the RegEx matches different strings, e.g. "goodbye world1 and world2" with pattern `(world\d)`? I think we would need to call the replacement function on each match individually in this case.

Reply:
Resolved [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29408/commits/7819437f67b6c511fb3a2f46bac74e485e6e9391)